### PR TITLE
[SPARK-46970][CORE] Rewrite `OpenHashSet#hasher` with `pattern matching`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
The proposed changes in this pr involve refactoring the method of creating a `Hasher[T]` instance in the code. The original code used a series of if-else statements to check the class type of `T` and create the corresponding `Hasher[T]` instance. The proposed change simplifies this process by using Scala's pattern matching feature. The new code is more concise and easier to read.

### Why are the changes needed?
The changes are needed for several reasons. Firstly, the use of pattern matching makes the code more idiomatic to Scala, which is beneficial for readability and maintainability. Secondly, the original code contains a comment about a bug in the Scala 2.9.x compiler that prevented the use of pattern matching in this context. However, Apache Spark 4.0 has switched to using Scala 2.13, and the new code has passed all tests, it appears that the bug no longer exists in the new version of Scala. Lastly, the change reduces the amount of code and makes it cleaner and easier to understand.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by LuciferYang** [comment]
Test first

File: core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala [hunk 1]
```diff
@@ -62,28 +62,12 @@ class OpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
...
-      (new FloatHasher).asInstanceOf[Hasher[T]]
-    } else {
-      new Hasher[T]
-    }
+  protected val hasher: Hasher[T] = classTag[T] match {
+    case ClassTag.Long => new LongHasher().asInstanceOf[Hasher[T]]
+    case ClassTag.Int => new IntHasher().asInstanceOf[Hasher[T]]
+    case ClassTag.Double => new DoubleHasher().asInstanceOf[Hasher[T]]
+    case ClassTag.Float => new FloatHasher().asInstanceOf[Hasher[T]]
+    case _ => new Hasher[T]
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala:70]: cc @cloud-fan @dongjoon-hyun @srowen 

All test passed. Can we make this change?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM, @LuciferYang .

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @LuciferYang .

**Comment by LuciferYang** [comment]
Thanks @dongjoon-hyun

**Comment by eheewag** [comment]
unsubscribe
Thanks,
Gabriel

**Pull request closed without merging**